### ISSUE TITLE
Suggestion: Add transaction reference id

### ISSUE THREAD

**Issue by jayprakashji** [nanocurrency/nano-node#2942]
**TLDR:**
Add int64 field which will serve as a payment identification between the 'merchant' and 'customer'.

**What?**
Currently in the fiat money world, merchant creates invoice number, which must be included in bank payment to pair it to invoice.
The reason is, that the merchant has one bank account for all customers, and needs to recognize the payment.
Even payment processors like Stripe, have optional parameters in payment api, where merchant can put invoice number, or database order id, for easier identification against his database.

**Why?**
There are many reasons, and I'm sure this list won't be definitive.
- If we want adaptation of Nano to real world, also Nano should adapt to the way whe world works and how people expect to work with transactions.
- All accounting software are expecting some payment referrence id, when pairing transactions to invoices. We can not expect, that the world changes accounting.
- Tax auditing expects, that you can prove, which payment was that.
- Ease of use for merchant and user. One recieving account, which can be added to adressbook, and referrence id in history to allways be able to check what was it I paid for.
- Payments are to the same recieving address, which can be included on official web of the company where everybody can check it, and put to adress book.
- can be used without payment speciffic QRcode
- would work (almost) the same way as bank payments currently, users in most situations would have easy transition to Nano
- Less ledger bloating ... counterintuitive, explained later

**Why in ledger and not outside (some api)?**
- ease of use. Nothing more is needed to identify payment.
- trust in data. Blocks are signed. Important, for example in the case of Tax auditing. Banks are trusted, because i.e. archiving laws. Signed ledger can not be changed, it is trusted by technology.
- history. In my bank account I can anytime check all payment details, without any other service. Users should expect the same from crypto payment.

**why int64?**
If we would allow string data, we would have to add 20-30 Bytes for it to be usable.
Users might missuse it, or put private data to permanent public ledger.
With unsigned int64, there are usable 19 digit numbers, which would be more than enough for any invoice number.
With int32, only 9 digits would be usable, and that is too few.
In binary form int64 requires 8Bytes, which should not be more than few percent added to trasnaction.

**Arguments against. (what I've heard during debates)**

_"We want instant transactions, anything added to transaction would slow it down."_
I don't know, how adding few bytes would make Nano slower.
Only thing I can think of is if it would overflow MTU and fragment packet, but I doubt that is possible with the data currently used. 

_"Unnecessary"_
Well, if some users do not need this, others do. Adding it won't affect the first group, but benefit the second.

_"World should change how it works to adapt to superior Nano"._
Companies and users won't use something, if they loose basic functionality and have to change how they work.
Nano should make users comfortable to adopt.

_"Direct block handoff option is planed to do this."_
Mentioned : https://forum.nano.org/t/work-generation-changes-for-consideration/914
and here : https://www.reddit.com/r/nanocurrency/comments/ii6pyc/discord_discussion_with_colin_this_morning/
"It involves delivering the block directly to the receiver, outside of the nano network, to know which payment is for which purchase."
It is too technical, far from simple.
It requires some api on merchant side, so it's unusable for some average joe who wants to accept nano.
It's still online QRcode-only payment, and it has to contain api address where to send the block.
etc...

_"Ledger bloating"_
I don't know how much one transaction takes in db, but I'd expect about 200-300B.
In completely basic thinking, 8B would be adding at most 4% to transaction size.

I don't have technical knowledge, how works the DB in Nano node, 
but most databases I know store data in blocks for adressing, and it would be low chance accident that 8B would overflow that block.
So adding 8B might not change the storage size.

And the biggest reason, is how payments work currently.

- Merchant creates new address ( open transaction )
- User pays to the new address ( send transaction + recieve transaction )
- Merchant transfers sum to from new address to his main account ( send transaction + recieve transaction )

That is far more ledger bloating, than 8Bytes added to send transaction.
And those new accounts can not be pruned. (you'd have to add some 'close account' block)

**PS:**
If you have too much free time, this was the initial reddit post https://www.reddit.com/r/nanocurrency/comments/ijbr59/transaction_message_content/

**Comment by qwahzi** [comment]
Hello jayprakashji! I think you might find some of the discussion [here](https://forum.nano.org/t/new-state-block-design/892) to be interesting. You might consider sharing your thoughts there as well :)
